File: async_http.go
package main

import (
	"fmt"
	"sync"
	"time"
)

var wg sync.WaitGroup

func HTTPGet(url string) chan []byte {
	fmt.Println("[HttpGet] http get1 start")
	done := make(chan []byte)

	go func() {
		defer wg.Done()
		time.Sleep(time.Second * 3)
		fmt.Printf("[HttpGet] http get %v done! \n", url)
		done <- []byte{'a', 'b', 'c'}
		fmt.Println("[HttpGet] http get end")
	}()

	return done
}

func HandlerResponse(url string) {
	go func() {
		defer wg.Done()
		r := <-HTTPGet(url)
		fmt.Println("[HandlerResponse] get the resp from HttpGet")
		fmt.Println("resp: ", string(r))
		fmt.Println("[HandlerResponse] handler resp done!");
	}()
}

func DoOtherWork() {
	fmt.Println("[otherWork] do some thing....")
}

func main() {
	wg.Add(2)

	HandlerResponse("www.baidu.com")
	DoOtherWork()

	wg.Wait()

	// Output:
	// [otherWork] do some thing....
	// [HttpGet] http get1 start
	// [HttpGet] http get www.baidu.com done! 
	// [HttpGet] http get end
	// [HandlerResponse] get the resp from HttpGet
	// resp:  abc
	// [HandlerResponse] handler resp done!
}

File: test/AsyncHTTP.cs
using System;
using System.Threading.Tasks;
using System.Threading;

namespace Test
{
    class Async1
    {
        // 异步任务，获取 url 的 resp 内容
        async Task<byte[]> HttpGet(string url) {
            Console.WriteLine("[HttpGet] http get start");
            byte[] resp = {1, 2, 3};
            await Task.Run(() => {
                Thread.Sleep(2000);
                Console.WriteLine("[HttpGet] http get {0} done!", url);
            });
            Console.WriteLine("[HttpGet] http get end");
            return resp;
        }

        // 对 resp 进行处理
        async void HandlerResponse(string url) {
            // 需要等待 HttpGet 执行完成
            var resp = await HttpGet(url);
            Console.WriteLine("[HandlerResponse] get the resp from HttpGet");
            Task.Run(() => {
                Console.WriteLine(resp);
                Console.WriteLine("[HandlerResponse] handler resp done!");
            });
        }

        void DoOtherWork() {
            Console.WriteLine("[otherWork] do some thing....");
        }

        static void Main(string[] args)
        {
            var a = new Async1();
            //Console.WriteLine("[main] thread start");

            // HandlerResponse 是一个异步任务
            a.HandlerResponse("www.baidu.com");
            
            //Console.WriteLine("[main] thread end");

            // 其他任务，这里不会被阻塞，因为上面的异步任务没有 await
            a.DoOtherWork();

            // 等待线程全部执行完成（不知道 C# 中如何优雅等待）
            Thread.Sleep(10000);
        }
    }
}